### ISSUE TITLE
An Indexed value is affected by a side effect after indexing

### ISSUE THREAD

**Issue by hatoo** [kenkoooo/optpy#100]
Thanks for fixing #58 

As you know, there is still an issue for `index` 

Here is some test code.
```python
def f(x):
    x[0] = 200
    return 0


x = [0]

print(x[0]+f(x))
```

python: 0
optpy: 200

> but I guess there are some ways to solve this problem, like separating Value::index to index and index_assign.

I think this would be a good option.

**Comment by kenkoooo** [comment]
Thank you for reporting it! I added this to the code base in #101.

**Comment by kenkoooo** [comment]
The issue was resolved in #102 by introducing `index_ref` and `index_value`. The test is enabled in #103.
